### PR TITLE
Fix: glob processing (fixes #11940)

### LINKED ISSUE #11940: Can't lint files with square brackets in the name
**Tell us about your environment**

MacOS 10.14.5 (18F132)
zsh 5.3 (x86_64-apple-darwin18.0)

* **ESLint Version:** v6.0.1
* **Node Version:** v8.15.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  "env": {
    "browser": true,
    "node": true,
    "es6": true,
  },
  "extends": [
    "eslint:recommended",
    "plugin:prettier/recommended",
  ],
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly",
  },
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module",
  },
  "rules": {
    "no-unused-vars": ["error", {args: "none"}],
  },
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

No source, the bug happens even if the file is empty.

```bash
$ cat src/routes/blog/\[slug\].svelte
console.log('x',)
$ npx eslint src/routes/blog/\[slug\].svelte
No files matching 'src/routes/blog/[slug].svelte' were found.
$ npx eslint src/routes/blog/[slug].svelte
No files matching 'src/routes/blog/[slug].svelte' were found.
$ npx eslint src/routes/blog/\\[slug\\].svelte
No files matching 'src/routes/blog/\[slug\].svelte' were found.
```

**What did you expect to happen?**

That eslint would lint the file.

**What actually happened? Please include the actual, raw output from ESLint.**

Eslint was unable to find the file.

**Are you willing to submit a pull request to fix this bug?**

I could give it a shot, but I'm probably not the best person to solve it.

### PR DESCRIPTION
**Pull request by mysticatea**
**What is the purpose of this pull request? (put an "X" next to item)**

[X] Bug fix: #11940

**What changes did you make? (Give an overview)**

This PR fixes glob processing to iterate target files. The new way is similar to how shell passes globs to programs. I.e., if there are files that the glob matched, it adopts the matched files. Otherwise, it adopts the glob string as is.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular.

### CODE REVIEW

File: tests/lib/cli-engine/cli-engine.js [hunk 1]
```diff
@@ -3454,6 +3454,46 @@ describe("CLIEngine", () => {
...
+                        "[ab].js": "",
+                        ".eslintrc.yml": "root: true"
+                    }
+                }).CLIEngine;
+                engine = new CLIEngine();
+
+                const { results } = engine.executeOnFiles(["[ab].js"]);
+                const filenames = results.map(r => path.basename(r.filePath));
+
+                assert.deepStrictEqual(filenames, ["a.js", "b.js"]);
```
Comment (aladdin-add) [inline tests/lib/cli-engine/cli-engine.js]: what if the user want to match "[ab].js" in this case?

just my 2c. :-) it seems more straightforward like this:
1. '[ab].js' as a filename
2. '[ab].js' as a glob pattern (if the file not exist).

Comment (mysticatea) [reply tests/lib/cli-engine/cli-engine.js]: At first, I thought so, but it's different from what shell does.

```
root@387125dc75aa:~# ls
[ab].js  a.js  b.js
root@387125dc75aa:~# echo [ab].js
a.js b.js
```

I think that it's better if the behavior is similar to popular tools.

Comment (aladdin-add) [reply tests/lib/cli-engine/cli-engine.js]: the question is the user cannot lint "[ab].js", even quoting `eslint "[ab].js"`.

I agreed we should be as close to popular tools as possible, but I'm not sure if we can make an exception here.

Comment (mysticatea) [reply tests/lib/cli-engine/cli-engine.js]: Yes. It's the problem that ESLint cannot distinguish `eslint "foo"` and `eslint foo`.

I'm considering that Windows has the same behavior with Linux with `eslint [ab].js`. Hmm.

**Comment by mysticatea** [comment]
OK. I updated to follow @aladdin-add 's suggestion.

`eslint [ab].js`-like commands have different behaviors between Windows and the others, it will be no problem because those are enough rare cases.

**Comment by aladdin-add** [comment]
just to clarify:  if the user quotes like `eslint "[ab].js"`, will have the same behavior between Windows and the others. am I reading correctly?

**Comment by mysticatea** [comment]
@aladdin-add Yes. `eslint "[ab].js"` has the same behavior on all platforms.

**Pull request merged**